Fix federated scope not shown when public addressbook upload is disabled

danxuliu opened the pull request:
Follow up to #26243

Besides a small change in a test, when the public address book is disabled now it is possible to select the _Federated_ scope for the profile attributes, as the address books can still be exchanged between trusted servers. Besides that the _Federated_ option is now hidden when the Federation app is disabled

Please note that this might need to be adjusted in the mobile apps as well. ~~Should [the property in the capabilities](https://github.com/nextcloud/server/pull/26243/commits/00d83a5db056a9c466831fd45347f414bd81f3ae#diff-583c5d160112290d35096fc2cd08f70ad41e9ef8cf7c0d108c10c67c0ae8854fR58) be renamed too?~~ The capability `AccountPropertyScopesFederationEnabled` was split in `AccountPropertyScopesFederatedEnabled` and `AccountPropertyScopesPublishedEnabled`.

## How to test (scenario 1)

- Log in as an admin
- Open the settings
- Open the _Sharing_ section under _Administration_
- Disable _Allow users to publish their data to a global and public address book_
- Open the _Personal info_ section
- Show the visibility menu on any of the elements, like _Phone number_

### Result with this pull request

The _Federated_ option is still available, although the _Published_ option is not

### Result without this pull request

Neither the _Federated_ nor the _Published_ options are available

## How to test (scenario 2)

- Log in as an admin
- Disable the _Federation_ app
- Open the settings
- Open the _Personal info_ section
- Show the visibility menu on any of the elements, like _Phone number_

### Result with this pull request

The _Federated_ option is no longer available

### Result without this pull request

The _Federated_ option is still available

## Additional context

It can be checked that even if uploads to the lookup server are disabled the address book is still exchanged between trusted servers in the following way:
- Start server A
- Create user A in server A
- Start server B
- Create user B in server B
- In server A, open the (Administration) Sharing section
- Disable _Allow users to publish their data to a global and public address book_
- In server B, open the (Administration) Sharing section
- Disable _Allow users to publish their data to a global and public address book_
- In server A, add server B as a trusted server
- In server B, add server A as a trusted server

Now both servers need to trust each other. This is done using background jobs, so either force somehow the jobs to run or wait patiently until they run :-) (or, if you are using the AJAX method, entertain yourself reloading pages on each server around 30 times, although you need to do it alternatively as the jobs need to run in both servers :party:). There are probably much better ways to do it, but in order to know when the servers have trusted each other you can print a log message in [TrustedServers::addSharedSecret](https://github.com/nextcloud/server/blob/cb057829f72c70e819f456edfadbb29d72dba832/apps/federation/lib/TrustedServers.php#L172)

Once the servers trust each other, synchronize their address books running `occ federation:sync-addressbooks` in both servers. If the sharing settings are then opened again the other server should be shown as trusted and synchronized.

Now, if you open the Contacts menu in any of the servers you will see the users from the other server.

danxuliu committed:
Add missing scope to test

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Fix "Federated" scope not shown when the lookup server is disabled

When the lookup server is disabled the address books can still be
exchanged between trusted servers. Therefore the user should be able to
set the "Federated" scope in that case.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu commented:
/backport to stable21

nickvergessen commented:
> Please note that this might need to be adjusted in the mobile apps as well. Should the property in the capabilities be renamed too?

Not renamed, but we actually need 2 capabilities?

danxuliu commented:
> Not renamed, but we actually need 2 capabilities?

I have checked and if the _federation_ app is disabled in server B it is no longer possible to synchronize its users from server A. So I guess that it would make sense to hide the _Federated_ scope if the _federation_ app is disabled.

However, if I am not mistaken in the WebUI the _Federated_ scope was never hidden before, so I do not know if there is a special reason for that or if it was just an oversight :shrug:

nickvergessen commented:
> However, if I am not mistaken in the WebUI the Federated scope was never hidden before, so I do not know if there is a special reason for that or if it was just an oversight shrug

Before @PVince81 recent work all scope settings were always hidden when the settings was off. Now it only hides the federation+published one as the others still make sense

blizzz commented:
So what's to do with this?

danxuliu committed:
Hide "federated" scope when Federation app is disabled

If the Federation app is disabled it is not possible to synchronize the
users from a different server.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Split capability

The "federated" and "published" scopes are independent one from each
other, so the capability that encompassed both needs to be split.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu commented:
Besides the initial fix the _Federated_ option is now hidden when the Federation app is disabled, and the capability was split in two different capabilities.

skjnldsv commented:
@danxuliu rebase? close?

danxuliu commented:
Rebased and ready to review again.

Pytal reviewed (approved):
~👍 will carry this over to Vue for 23 :)~

Vuetification of the Personal info settings page has been postponed, should get this in :)
